pss/trojan/message.go

```diff
@@ -108,6 +111,31 @@ func (m *Message) Wrap(targets [][]byte) (chunk.Chunk, error) {
 	return chunk, nil
 }
 
+// Unwrap attemps to determine whether the given chunk is a trojan chunk
+// it will return the resulting message if the unwrapping is successful, and an error otherwise
+// TODO: move this to PSS API?
+func Unwrap(c chunk.Chunk) (*Message, error) {
+	d := c.Data()
+
+	// unmarshal chunk payload into message
+	m := new(Message)
+	// first 40 bytes are span + nonce
+	if err := m.UnmarshalBinary(d[40:]); err != nil {
+		return nil, err
+	}
+
+	// check for message correctness
```

Comment:
? why are you checking the localstore checks it

Reply:
as i understand it:
1- this func will be called by the localstore (on the receiving end of the trojan chunk)
2- if we don't hash and check for correctness, any chunk could have its payload marshalled into a nonce + a trojan message, as they are simply byte arrays.

Reply:
3 is taken care of by the localstore content address chunk validator.

whether it makes sense as a trojan,is taken care of by unmarshalling
and potentially an integrity check

Reply:
ok so i'm a bit confused: as i see it, "unmarshalling and an integrity check" is precisely what is going on here.

if this is already happening somewhere else in the code, can you please point out where? 

i am aware of the `Validate` func from `storage/feed/handler.go:83`, but afaik it does not check for trojan correctness (should it?).

my worry here is that if the `check for message correctness` block is removed from `Unwrap`, the function could be called on any valid non-trojan chunk and would create a garbage trojan message.

now: if this won't happen because a validation would be called before `Unwrap` takes place, that's perfectly fine. i would just like to see where the trojan content is verified to be correct before this occurs.

hope that's clear

Reply:
ah ok, so the validation is outside the localstore now, for some reason: 
- inserted here; https://github.com/ethersphere/swarm/blob/master/swarm.go#L237
- called here: https://github.com/ethersphere/swarm/blob/master/chunk/chunk.go#L307

the trojan hook should come after this validation. 
an integrity check plus extra obfuscation of the topic can be done if the topic is encrypted with a key that is 

```
key := Hash(topic, payload)
```

now the integrity check would be:

```golang
for topic, handler := range handlers {
     if xor(hash(topic, m.payload),topic) == m.topic {
         go handler(m)
         break
     }
}
```

Reply:
thanks for clearing this up.

i've removed the `check for message correctness` code and removed  the `TestUnwrapFail` func as well—makes no sense to have it without said block of code.

nevertheless i've added a comment to the function so that the assumptions we're making here are clear. i will also be on the lookout to make sure the right validator is called in the final implementation for this feature.

i will leave the topic checking for a later iteration if we really need it (created issue #2160 for this).